enhancements/cluster-logging/loki-gateway-configuration.md

```diff
@@ -0,0 +1,347 @@
+---
+title: loki-lokistack-gateway-configuration
+
+authors:
+  - "@sasagarw"
+
+reviewers:
+  - "@periklis"
+
+approvers:
+  - "@periklis"
+  - "@igor-karpukhin"
+
+creation-date: 2021-08-27
+last-updated: 2021-08-27
+status: implementable
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Loki Operator: LokiStack CR extension for the gateway configuration
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+LokiStack Gateway is a component deployed as part of Loki Operator. It secure access on the loki cluster distributor and query-frontend endpoint via an OAuth/OIDC-handling gateway.
+
+This proposal seeks to provide OIDC configuration for the LokiStack gateway to integrate Loki access to the org's OIDC provider.
+
+## Motivation
+
+### Goals
+
+The goals of this proposal are:
+
+* Provide changes to API to accept OIDC provider in LokiStack CR
+* Make OIDC configuration optional and auto-configuration on OCP
+* Secrets/Credentials/SA-Keys should be provided via a k8s secret that the LokiStack admin provides upfront
+
+We will be successful when our LokiStack CR is able to accept OIDC provider and able to configure LokiStack Gateway
+
+### Non-Goals
+
+## Proposal
+
+For configuring the LokiStack Gateway, we need information about oidc provider and opa.
+
+### lokistack_types.go
+
+Add a new field `GatewayConfiguration` to `LokiTemplateSpec` struct as below:
+
+```go
+// LokiTemplateSpec defines the template of all requirements to configure
+// scheduling of all Loki components to be deployed.
+type LokiTemplateSpec struct {
+    // GatewayConfiguration defines the lokistack-gateway component configuration spec.
+	//
+	// +optional
+	// +kubebuilder:validation:Optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Gateway Configuration"
+	GatewayConfiguration *LokiGatewayConfigurationSpec `json:"gatewayConfiguration,omitempty"`
```

Comment:
The `LokiTemplateSpec` struct consists of set of fields where each represents a component. `GatewayConfiguration` as field does not represent a component but rather extra configuration for the `Gateway` component. I suggest to use struct embedding here to make the API cleaner, i.e.:
```go
type LokiTemplateSpec struct {
    Gateway *GatewayTemplateSpec
}

type GatewayTemplateSpec struct {
    LokiComponentSpec
    Tenants []TenantSpec
    Rbac *RbacSpec
}
```

An alternative here is to keep `Gateway` as `*LokiComponentSpec` because `LokiTemplateSpec` is all about the deployment/scheduling of component pods on k8s. The tenants/rbac mix could live in a separate struct referenced in the toplevel LokiStackSpec, e.g.:
```go
type TenantsSpec struct {
    Tenants []TenantsSpec `json:"tenants,omitempty"`
    Rbac *RbacSpec `json:"rbac,omitempty"`
}

type LokiStackSpec struct {
    Tenants *TenantsSpec
}
```